hank-client/src/main/java/com/liveramp/hank/client/HostConnectionPool.java

```diff
@@ -243,90 +264,88 @@ private int getNextHostIndexToUse(int previouslyUsedHostIndex) {
+
     while (true) {
+
+      //  jump out if we don't have any more preferred hosts
+      if (numPreferredTries >= preferredPools.hostToConnections.size()) {
+        break;
+      }
+
       // Either get a connection to an arbitrary host, or get a connection skipping the
       // previous host used (since it failed)
-      if (connectionAndHostIndex == null) {
-        if (keyHash == null) {
-          connectionAndHostIndex = getConnectionToUse();
-        } else {
-          connectionAndHostIndex = getConnectionToUseForKey(keyHash);
-        }
-      } else {
-        connectionAndHostIndex = getNextConnectionToUse(connectionAndHostIndex.hostIndex);
+      connectionAndHostIndex = getConnectionFromPools(preferredPools, keyHash, connectionAndHostIndex);
+
+      ++numPreferredTries;
+
+      HankResponse response = attemptQuery(connectionAndHostIndex, domain, key, numPreferredTries, maxNumTries);
+      if (response != null) {
+        return response;
       }
-      // If we couldn't find any available connection, return corresponding error response
-      if (connectionAndHostIndex == null) {
-        LOG.error("No connection is available. Giving up. Domain = " + domain.getName() + ", Key=" + BytesUtils.bytesToHexString(key));
-        return NO_CONNECTION_AVAILABLE_RESPONSE;
-      } else {
-        // Perform query
-        try {
-          return connectionAndHostIndex.hostConnection.get(domainId, key);
-        } catch (IOException e) {
-          // In case of error, keep count of the number of times we retry
-          ++numTries;
-          if (numTries < maxNumTries) {
-            // Simply log the error and retry
-            LOG.error("Failed to perform query with host: "
-                + connectionAndHostIndex.hostConnection.getHost().getAddress()
-                + ". Retrying. Try " + numTries + "/" + maxNumTries
-                + ", Domain = " + domain.getName()
-                + ", Key = " + BytesUtils.bytesToHexString(key), e);
-          } else {
-            // If we have exhausted tries, return an exception response
-            LOG.error("Failed to perform query with host: "
-                + connectionAndHostIndex.hostConnection.getHost().getAddress()
-                + ". Giving up. Try " + numTries + "/" + maxNumTries
-                + ", Domain = " + domain.getName()
-                + ", Key = " + BytesUtils.bytesToHexString(key), e);
-            return HankResponse.xception(HankException.failed_retries(maxNumTries));
-          }
-        }
+    }
+
+    while (true) {
+      connectionAndHostIndex = getConnectionFromPools(otherPools, keyHash, connectionAndHostIndex);
+      ++numOtherTries;
+
+      HankResponse response = attemptQuery(connectionAndHostIndex, domain, key, numPreferredTries+numOtherTries, maxNumTries);
+      if (response != null) {
+        return response;
       }
     }
+
   }
 
-  public HankBulkResponse getBulk(int domainId, List<ByteBuffer> keys, int maxNumTries) {
```

Comment:
Nothing used it anymore so I wasn't inclined to rewrite and test it.  Bulks seem to be handled entirely at the HankSmartClient level using Futures now, so I think this was just vestigial.